### PR TITLE
accounts/keystore: fix cache update and TestUpdatedKeyfileContents fa…

### PR DESCRIPTION
**Pull request by gbotrel**
fix cache update and TestUpdatedKeyfileContents failing on macOS.
TestUpdatedKeyfileContents was randomly failing on macOS. Root cause: after writing to a file and closing the file handle, we get the fs notify event, we scan the directory and the modified time of the file is not yet updated (and the scanAccount logic relies on that to update the cache), even after the 500ms delay in watch.go. 
I believe using the EventInfo from notify library in watch.go:93 to avoid scanning directory and relying on modified time of files would be safer (the event contains the Modified / Created / Deleted file path, yet we scan the whole directory and compare modified times)

### CODE REVIEW

File: accounts/keystore/account_cache.go [hunk 1]
```diff
@@ -260,7 +261,8 @@ func (fc *fileCache) scanFiles(keyDir string) (set.Interface, set.Interface, set
 			continue
 		}
 		filesNow.Add(path)
-		if modTime.After(prevMtime) {
+		// on macOS, we get FS notifications before the actual modTime is updated
+		if modTime.After(prevMtime) || (modTime.Equal(prevMtime) && runtime.GOOS == "darwin") {
```
Comment (fjl) [inline accounts/keystore/account_cache.go]: Maybe a better fix would be to use `modTime.After(prevMtime) || modTime.Equal(prevMtime)` on all platforms.
Does that sound good to you?

Comment (gbotrel) [reply accounts/keystore/account_cache.go]: Yes if we want to avoid platform specific code, I believe the perf impact is negligible

**Comment by holiman** [comment]
This means rescanning the entire directory again, which is one of the basic problems I wanted to avoid. It is non-neglible when there are hundreds of thousands of accounts. So if there's any other way of doing it (e.g differebt lib), I'd prefer that. Alternatively, keep this fix as a mac only solution. My 5 c.

**Comment by gbotrel** [comment]
Does it? (= rescan the entire directory). My understanding is that this scan is now triggered only after a FS event is thrown. Meaning something did effectively change in the directory --> meaning iteration i+2 will likely update the new reference modified time: 
`if modTime.After(newMtime) {
	newMtime = modTime
}`

and all the files not changing there modTime won't be picked again.

This was a quick fix, I agree there are more elegant solutions (see first comment --> what about using EventInfo data that contains the path of the file that changed ? but that does change much more than just a line of code)

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: d51e32a0db2e88d55b8b24e9607a732ff4ad6d58
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: d51e32a0db2e88d55b8b24e9607a732ff4ad6d58
  * Commits must be prefixed with the package(s) they modify

 * Commit: c7f853db290b5500abaf8c073ee011e3fc83e1fb
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: d51e32a0db2e88d55b8b24e9607a732ff4ad6d58
  * Commits must be prefixed with the package(s) they modify

 * Commit: c7f853db290b5500abaf8c073ee011e3fc83e1fb
  * Commits must be prefixed with the package(s) they modify

 * Commit: b809a9066938b58b7bda26c2ddde52e224a1cb55
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by fjl** [comment]
Please remove travis-related changes from this PR.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: d51e32a0db2e88d55b8b24e9607a732ff4ad6d58
  * Commits must be prefixed with the package(s) they modify

 * Commit: c7f853db290b5500abaf8c073ee011e3fc83e1fb
  * Commits must be prefixed with the package(s) they modify

 * Commit: b809a9066938b58b7bda26c2ddde52e224a1cb55
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5736f0269721ac52bd1c961ec8609b21d6722aa9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: d51e32a0db2e88d55b8b24e9607a732ff4ad6d58
  * Commits must be prefixed with the package(s) they modify

 * Commit: c7f853db290b5500abaf8c073ee011e3fc83e1fb
  * Commits must be prefixed with the package(s) they modify

 * Commit: b809a9066938b58b7bda26c2ddde52e224a1cb55
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5736f0269721ac52bd1c961ec8609b21d6722aa9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by gbotrel** [comment]
Did remove the guilty commits with a history rewrite, please ignore GitCop messages :-)

**Comment by holiman** [comment]
> Does it? (= rescan the entire directory). My understanding is that this scan is now triggered only after a FS event is thrown. Meaning something did effectively change in the directory --> meaning iteration i+2 will likely update the new reference modified time:
if modTime.After(newMtime) { newMtime = modTime }

>and all the files not changing their modTime won't be picked again.

So, the `newMtime` gets set to whatever file was most recently updated, of all files. So you're right, it will not rescan all of them, just the one(s) that have the latest modiifcation time. 

However, that makes me question if this fix will work. Consider the following: 

1. There are a bunch of files, A (mtime:1), B (mtime: 2) and C (mtime: 3). 
2. A scan is made. `fc.mtime` becomes set to `3` (meaning). 
3. `A` is modified, and a rescan is made before the `mtime` of `A` is updated. 
4. The file `C` will be read again, but not `A`. 

Or is there another fs-event triggered once the `mtime` is changed?

**Comment by holiman** [comment]
This one: https://github.com/ethereum/go-ethereum/pull/15498.

**Comment by gbotrel** [comment]
The other PR don't seem to fix the underlying problem --> the test will pass, but the scenario we mentioned here can still happen and result in a cache not being updated. I agree with your previous comment, my fix is not enough for this to be bulletproof, what about using the EventInfo from notify library? If I have some time will run some tests the coming days.

**Comment by holiman** [comment]
I'll make a PR using the EventInfo thing, and hopefully the buildsystem can tell us how cross-platform portable it is.

**Comment by holiman** [comment]
@gbotrel I can't reproduce the issues locally, but if you're on a Mac, you're more than welcome to test my PR (linked above) and see if it fixes the issue for you.

**Comment by gbotrel** [comment]
@holiman  --> did test your PR #15580 on mac, runs fine 👍 . But it has the other PR (#15498 ) changes in it, making the test quite longer than it was. 
I did remove the multiple sleeps
```
// needed so that modTime of `file` is different to its current value after ioutil.WriteFile
// time.Sleep(1000 * time.Millisecond)
```

And the tests pass. I believe adding the sleeps was just hiding the test failure, and your PR solves the actual underlying problem. 

Closing this PR .

**Comment by holiman** [comment]
Great, thanks for reporting! I'll change those values back, and clean up the PR, and mark it as no-longer-wip.

**Pull request closed without merging**